graph-commons/src/main/scala/io/renku/triplesstore/LuceneQueryEncoder.scala

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2022 Swiss Data Science Center (SDSC)
+ * A partnership between École Polytechnique Fédérale de Lausanne (EPFL) and
+ * Eidgenössische Technische Hochschule Zürich (ETHZ).
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.renku.triplesstore
+
+object LuceneQueryEncoder {
+
+  private val specialCharactersEscape =
+    List("+",
+         "-",
+         "&",
+         "&&",
+         "|",
+         "||",
+         "!",
+         "(",
+         ")",
+         "{",
+         "}",
+         "[",
+         "]",
+         "^",
+         """"""",
+         "~",
+         "*",
+         "?",
+         ":",
+         """\""",
+         "/"
+    ).map {
+      case c if c == "}" || c == "]" || c == "*" || c == """\""" => c -> s"""\\$c"""
+      case c                                                     => c -> s"""\\\\$c"""
+    }.toMap
+
+  def queryAsString(v: String): String =
+    specialCharactersEscape.foldLeft(v) { case (escaped, (specialChar, replacement)) =>
```

Comment:
How does it work for strings like `"a&&b"` wrt the `&` vs `&&` replacements (same for `||`)? Wouldn't it be first replaced to `a\&\&b` so that the rule for `&&` would never trigger? We could perhaps use a list of tuples with sorting by rule preferences. Or it looks like we don't even need the `&&` and `||` rules? (then we could also use a `Char` type :))

Reply:
OTOH maybe it worth thinking about using the java lucene library to let it escape these strings?